runtime/src/iree/builtins/ukernel/pack.c

```diff
@@ -0,0 +1,91 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/builtins/ukernel/pack.h"
+
+static iree_ukernel_status_t iree_ukernel_pack_validate(
+    const iree_ukernel_pack_params_t* params) {
+  const uint32_t allflags = IREE_UKERNEL_FLAG_PACK_TRANSPOSE_INNER |
+                            IREE_UKERNEL_FLAG_PACK_TRANSPOSE_OUTER;
+  if (params->flags & ~allflags) {
+    return iree_ukernel_status_bad_flags;
+  }
+  switch (params->type) {
+    case iree_ukernel_pack_type_f32f32:
+    case iree_ukernel_pack_type_i8i8:
+    case iree_ukernel_pack_type_i32i32:
+      break;
+    default:
+      return iree_ukernel_status_bad_type;
+  }
+  if (params->in_stride0 < 0 || params->out_stride0 < 0 ||
+      params->in_size0 < 0 || params->in_size1 < 0 || params->out_size0 < 0 ||
+      params->out_size1 < 0 || params->out_size2 < 0 || params->out_size3 < 0) {
```

Comment:
So far in ukernels (for the existing `mmt4d` ukernel) I had assumed that `0` was a valid size (and resulted in a no-operation, but still valid). But we shouldn't have do decide that here; what does `linalg`, `linalg_ext` say? Note, this is not necessarily about the validity of static shapes such as `5x0xi32`. This could be a dynamic shape such as `?x?xi32` that happens to have a value `0` at runtime.  I had defaulted to assuming that `0` would be a legal no-op case out of experience with various other matrix/array computation libraries, where it's typically treated that way, and in my experience, the reason why it has to be treated that way is that it can be annoying when `0` is an error case, it can force callers to special-case it on their side.

Reply:
I see, thanks for explanation! I was just curious about the check. I remember that there are some discussion about tensor<0xi32> stuff, but I don't remember details. They always look weird to me. :p

Reply:
https://reviews.llvm.org/D85413 introduces semantics of tensor<0xf32> and memref<0xf32> for Linalg ops. It looks like tensor<0xf32> is expected to be lowered to a scalar. I think we don't have to consider the semantics in linalg because we probably wont have real cases like this.

Reply:
!!!!! tensor<0xf32> is a scalar?!

It would violate so many invariants, if an operation on a tensor<0xf32> was an operation on 1 scalar. It should be an operation on 0 scalars, i.e. a no-op.

Reply:
I looked a bit into https://reviews.llvm.org/D85413 . It distinguishes between `memref<0xf32>` and `tensor<0xf32>`. It says that the former can be DCE'd, and even guards that in a lit test in `mlir/test/Dialect/Linalg/canonicalize.mlir`. So as far as `memref` is concerned, it agrees with us that those should be no-ops. That's enough for our immediate purposes here, since the input dialect that we are lowering here (in `LowerLinalgMicrokernels`) is linalg-on-memref. We never deal with tensors in this PR.

Separately, https://reviews.llvm.org/D85413 says that `tensor<0xf32>` cannot be DCE'd and needs to be lowered to `f32` or some `memref` containing a single `f32`, but that doesn't necessarily imply that it literally means for this artificial `f32` to survive the entire compilation --- maybe it just means that it needs to be kept artificially just a little longer, but it still goes away? Anyway, this tensor-specific discussion is not directly relevant here as we are only dealing with memref.